src/write/cfi.rs

```diff
@@ -0,0 +1,909 @@
+                augmentation_length += 1 + u64::from(encoding.address_size);
+            }
+            // TODO: R (FDE address encoding)
+            if self.signal_trampoline {
+                w.write_u8(b'S')?;
+            }
+        }
+        w.write_u8(0)?;
+
+        if encoding.version >= 4 {
+            w.write_u8(encoding.address_size)?;
+            // TODO: segment_selector_size
+            w.write_u8(0)?;
+        }
+
+        w.write_uleb128(self.code_alignment_factor.into())?;
+        w.write_sleb128(self.data_alignment_factor.into())?;
+
+        // TODO: eh_frame encoding
+        if encoding.version == 1 {
+            let register = self.return_address_register.0 as u8;
+            if u16::from(register) != self.return_address_register.0 {
+                return Err(Error::ValueTooLarge);
+            }
+            w.write_u8(register)?;
+        } else {
+            w.write_uleb128(self.return_address_register.0.into())?;
+        }
+
+        if augmentation {
+            w.write_uleb128(augmentation_length)?;
+            let offset = w.len();
+            if self.lsda {
+                // TODO: allow encoding to be specified
+                w.write_u8(constants::DW_EH_PE_absptr.0)?;
+            }
+            if let Some(address) = self.personality {
+                // TODO: allow encoding to be specified
+                w.write_u8(constants::DW_EH_PE_absptr.0)?;
+                w.write_address(address, encoding.address_size)?;
+            }
+            // TODO: R (FDE address encoding)
+            debug_assert_eq!(w.len(), offset + augmentation_length as usize);
+        }
+
+        for instruction in &self.instructions {
+            instruction.write(w, self)?;
+        }
+
+        write_nop(
+            w,
+            encoding.format.word_size() as usize + w.len() - length_base,
+            encoding.address_size,
+        )?;
+
+        let length = (w.len() - length_base) as u64;
+        w.write_initial_length_at(length_offset, length, encoding.format)?;
+
+        Ok(offset)
+    }
+}
+
+/// A frame description entry. There should be one FDE per function.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct FrameDescriptionEntry {
+    /// The initial address of the function.
+    address: Address,
+
+    /// The length in bytes of the function.
+    length: u32,
+
+    /// The address of the LSDA.
+    pub lsda: Option<Address>,
+
+    /// The instructions for this function, ordered by offset.
+    instructions: Vec<(u32, CallFrameInstruction)>,
+}
+
+impl FrameDescriptionEntry {
+    /// Create a new frame description entry for a function.
+    pub fn new(address: Address, length: u32) -> Self {
+        FrameDescriptionEntry {
+            address,
+            length,
+            lsda: None,
+            instructions: Vec::new(),
+        }
+    }
+
+    /// Add an instruction.
+    ///
+    /// Instructions must be added in increasing order of offset, or writing will fail.
```

Comment:
If we internally stored instructions as `BTreeMap<u32, Vec<CallFrameInstruction>>` the user-facing API could be a bit nicer here. We wouldn't have to have the order of adding instructions requirement, and `write_advance_loc` would also get a little simpler.

Reply:
Lots of small Vecs seems bad for performance though?

Reply:
Also, I don't think the ordering requirement is problematic, because the machine code and line number program need to be emitted in order too, and the CFA definitions depend on order (although maybe we could handle that internally too).

Reply:
Fair